src/extension/app/store/bulk.js

```diff
@@ -0,0 +1,652 @@
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import { action, observable } from 'mobx';
+import { log } from '../../log.js';
+import {
+  EXTERNAL_EVENTS,
+  MODALS,
+  MODAL_EVENTS,
+  STATE,
+} from '../constants.js';
+
+/**
+ * @typedef {import('./app.js').AppStore} AppStore
+ */
+
+/**
+ * @typedef {import('@Types').BulkResource} BulkResource
+ */
+
+/**
+ * @typedef {import('@Types').BulkSelection} BulkSelection
+ */
+
+/**
+ * @typedef {import('@Types').BulkSummary} BulkSummary
+ */
+
+/**
+ * @typedef {import('@Types').AdminJob} AdminJob
+ */
+
+/**
+ * @typedef {import('@Types').AdminJobProgress} AdminJobProgress
+ */
+
+/**
+ * @typedef {import('@Types').AdminJobResource} AdminJobResource
+ */
+
+/**
+ * The path prefix for illegal file names.
+ * @type {string}
+ */
+const illegalPathPrefix = '!ILLEGAL!_';
+
+/**
+ * The file types in Google Drive. IDs represent the last 4 characters
+ * of first path in the corresponding SVG icons.
+ * @type {Object<string, string>}
+ */
+export const gdriveFileTypes = {
+  folder: '2-2z',
+  sharedFolder: '2v1z',
+  gdoc: '777z',
+  gsheet: '778z',
+  word: '888z',
+  excel: '444z',
+  image: '1-4z',
+  svg: '1-4z',
+  pdf: '666z',
+  video: '3.2z',
+};
+
+/**
+ * Handles bulk operations.
+ */
+export class BulkStore {
+  /**
+   * The bulk selection
+   * @type {BulkSelection}
+   */
+  @observable accessor selection = [];
+
+  /**
+   * The bulk progress
+   * @type {AdminJobProgress}
+   */
+  @observable accessor progress = null;
+
+  /**
+   * @type {AppStore}
+   */
+  appStore;
+
+  /**
+   * @type {BulkSummary}
+   */
+  summary;
+
+  @action
+  setSelection(selection) {
+    this.selection = selection;
+  }
+
+  /**
+   * Validates a bulk resource.
+   * @param {BulkResource} item The bulk resource
+   * @returns {BulkResource} The validated bulk resource
+   */
+  #validateBulkResource(item) {
+    // detect illegal characters in file name
+    const { file, type } = item;
+    if (['/', '*', '\\', '!', '?'].find((char) => file.includes(char))) {
+      return {
+        type,
+        file: `${illegalPathPrefix}${file}`,
+      };
+    }
+
+    return {
+      type,
+      file,
+    };
+  }
+
+  /**
+   * Normalizes a file name.
+   * @param {string} file The file name
+   * @returns {string} The sanitized file name
+   */
+  #normalize(file) {
+    return file
+      .toLowerCase()
+      .normalize('NFD')
+      .replace(/[\u0300-\u036f]/g, '')
+      .replace(/[^a-z0-9]+/g, '-')
+      .replace(/^-|-$/g, '');
```

Comment:
Do we want to do anything about file names that will end up being empty? I.E $.docx

Reply:
Maybe, but I didn't want to deviate from the server side normalization here. Currently it's 1:1.